### PR TITLE
doc: Add CODEOWNERS file to automatically nominate PR reviewers

### PR DESCRIPTION
**Pull request by adamjonas**
This PR brings back and builds on #17094. GitHub uses a CODEOWNERS magic file to automatically add tagged contributors to the "Reviewers" list for a PR.

The goal of this is to make use of GitHub's suggested reviewers feature and not to confer ownership or give veto power to specific people. It would be better if this file could be named CODEREVIEWERS, but alas, that wouldn't work. The idea of a NAGFILE was proposed in [Bitcoin Core Dev meeting in 2018](https://diyhpl.us/wiki/transcripts/bitcoin-core-dev-tech/2018-03-07-priorities/#:~:text=NAGFILE). While this GitHub implementation has some complications, it's a step towards realizing the promise of automating "reviewing begging" and (hopefully) positively impacting the review process as a whole.

Of secondary value, this file can serve as documentation for who the maintainers are and who it might be smart to check with for certain areas of code/features (i.e., fuzzing, PSBT, and Bech32) -- this is helpful information for new contributors.

* The first commit is taken from #17094
* The second commit adds comments and expands the list of reviewers based on the suggestions and comments from that PR
* ~The third WIP commit~ This commit also uses the doc dir as an example of granular assignments based on lines of codes ~contributed~ written and/or general engagement with the project. (If interested, here is a report for [most lines of code per author for each file](https://gist.github.com/adamjonas/854a46a1918224927b186865baeac411)). The pro of this level of detail is that the best reviewer is more likely to be nominated. The con is that it will create churn as files are renamed, new files are added, or reviewers want to be added or removed.

Some open questions:
* How often should this file be changed?
* What level of history does one need have on the project before being added to this file? When does it make sense to remove a reviewer?
* These review notifications can [cause a lot of noise](https://github.community/t5/How-to-use-Git-and-GitHub/Team-based-notifications-or-rework-CODEOWNERS-notification/td-p/7811) and automatically subscribes the requested reviewer to the thread. A GitHub Team based approach would allow for adding or removing reviewers without modifying this file; however, this comes along with its [own set of problems](https://bionic.fullstory.com/taming-github-codeowners-with-bots/#problems-with-github-code-owners), including granting [write access](https://github.community/t5/How-to-use-Git-and-GitHub/CODEOWNERS-works-with-users-but-not-teams/td-p/4986#U4991). Other projects [have used bots](https://bionic.fullstory.com/taming-github-codeowners-with-bots/#using-a-github-bot) to sidestep this.

### CODE REVIEW

**Comment by elichai** [comment]
I think what @MarcoFalke suggested here: https://github.com/bitcoin/bitcoin/pull/17094#issuecomment-561194286 is a good idea,
this can be a way also for contributors to subscribe to PRs touching specific parts of the code,
that way they can know if something might conflict with their PRs or they can know when there's something they feel comfortable reviewing.

I do wonder though how it will look like when each folder might have a long list of people,
Although if it's a problem we maybe could use githubs teams feature, and create teams like `validation`, `crypto`, `p2p`, `wallet` etc. and subscribe those teams and then add people to the teams accordingly, but maybe that's an overkill.

File: CODEOWNERS
Comment (jonatack) [inline CODEOWNERS:97]: The rapidcheck documentation was recently removed along with rapidcheck itself.

Comment (adamjonas) [reply CODEOWNERS:97]: Ah, thanks. This section currently serves as an example of what is possible, but your comment also illustrates the point in the PR description that the downside of a detailed CODEOWNERS file is that it will require more attention and updates.

**Comment by jonatack** [comment]
Could be worth trying but I'll defer to the maintainers. IIUC this would be a guideline/suggestion and not a limiter, but I'm not sure LOC "contributed" is great metric. Review time spent on the codebase and willingness to invest time in a PR may both be considerations as well. Also, PRs can be improved by repeated rounds of reviewing but in the end only the PR author is taken into account in the LOC stats. Question also if the updating and maintainance effort would be worth the value added.

**Comment by jonatack** [comment]
Could be worth trying but I'll defer to the maintainers. IIUC this would be a guideline/suggestion and not a limiter, but I'm not sure LOC "contributed" is a great metric. Review time spent on the codebase and willingness to invest time reviewing a particular PR may both be factors as well, but which would not be accounted for. Also, PRs can be improved by repeated rounds of reviewing but in the end only the PR author is taken into account in the LOC stats. It may be good for these factors to be stated if this file is intended to be a reference for new contributors. A final question is if the updating and maintainance effort would be worth the value added.

File: CODEOWNERS
Comment (jonatack) [reply CODEOWNERS:97]: > your comment also illustrates the point in the PR description that the downside of a detailed CODEOWNERS file is that it will require more attention and updates.

Agreed. It may be good for someone to be nominated/dedicated to this.

**Review by ryanofsky — APPROVED** [review]
Code review ACK cfd2792a375025b05508fe954e5c0cdfcf41a666. This looks great, and very thoughtfully implemented. I'd like to be added (in this PR or after) as a reviewer for src/interfaces/, and probably a few other places too like src/util/settings/.

On the concept, I think it's a potentially promising experiment. We'll have to see how it works and how people change the file and use the feature to match their preferences. Maybe it will even encourage less monolithic code organization and breaking up giant files like wallet.cpp into modules that people want to receive review requests for. In past projects, I've found systems for tracking review requests very useful, not just as a guide for what to review next, but also for requesting reviewers I might not have thought of, and for steering new requests away from people who are already overloaded with too many existing ones

**Review by ariard — COMMENTED** [review]
Concept ACK, I think that's nice for newcomers, specially with regards to less visible part of the project like build or doc. Though I do think that name CODEOWNERS is pretty deceiving.

I do like Elichai idea of notifications list to subscribe by labels.

File: CODEOWNERS
Comment (fjahr) [inline CODEOWNERS:80]: I think this section should be in the front. If a PR touches code and one of these files I think the code reviewers should be the ones getting tagged. Meaning I think most of the time that is the better fit rather than the other way around.

**Review by fjahr — COMMENTED** [review]
Concept ACK

We can't be sure how this will work out long term but I think this is worth a try. My biggest concern with this would be that a reviewer that is added to a PR might be a bottleneck. There just needs to be clarity that maintainers don't need to wait for their reviews when something is RFM and if the automatically added reviewers don't have the time or feel they are not the right person they remove themselves. So good job on clarifying what this is and what it isn't!

**Comment by laanwj** [comment]
Concept ACK

File: CODEOWNERS
Comment (laanwj) [inline CODEOWNERS:91]: I'm not sure assigning me for the windows build doc is a good idea :smile:

File: CODEOWNERS
Comment (adamjonas) [reply CODEOWNERS:13]: I like the spirit of this but updated with my own wording.

File: CODEOWNERS
Comment (adamjonas) [reply CODEOWNERS]: Makes sense. Moved.

File: CODEOWNERS
Comment (adamjonas) [reply CODEOWNERS]: Substituted sipsorcery instead.

**Comment by MarcoFalke** [comment]
Concept ACK, I think there is nothing to lose trying this out for a few weeks or months and keeping it if it works.

Slightly stronger Concept ACK, if my name was removed from the list, since I will get notifications either way.

**Review by ryanofsky — APPROVED** [review]
Code review ACK cfd604a1a263e80f88596c043c41dc6b15dd4c9a. Changes since previous review: names, dates, places, things. So much

**Comment by fanquake** [comment]
> Slightly stronger Concept ACK, if my name was removed from the list, since I will get notifications either way.

Same. I already get emails for everything that happens on this repo. So can probably skip this if it's going to add duplicates.

**Comment by ryanofsky** [comment]
> > Slightly stronger Concept ACK, if my name was removed from the list, since I will get notifications either way.
> 
> Same. I already get emails for everything that happens on this repo. So can probably skip this if it's going to add duplicates.

I'd be really surprised if github started sending duplicate notifications. I think listing names here could not just be useful for automated emails, but also for project documentation, to give people an idea of who they could ping if they need help with an issue or pull request. Of course if you don't want to be contacted or receive notifications your name shouldn't be here, but it sounds like that's a different issue?

**Comment by fanquake** [comment]
> I'd be really surprised if github started sending duplicate notifications

I thought this was the case last time I received a review request, but the behaviour may have changed. I've created a [repo here](https://github.com/fanquake/gh_notif_test) if someone wants to open a PR and test the behaviour.

> but also for project documentation, to give people an idea of who they could ping if they need help with an issue or pull request.

In that case also put me down for symbol/security-check scripts/tests and glibc compat.

**Comment by fanquake** [comment]
> I thought this was the case last time I received a review request, but the behaviour may have changed.

Looks like this is still the case; you'll get an email for the PR, and then a followup with the review request (thanks @elichai):

emails

However I think I care about this less now. Feel free to add me for the files I mentioned above.

**Comment by MarcoFalke** [comment]
ACK 375dcd15566213c4d58e05a9738c8b5b1360a0d1

**Comment by adamjonas** [comment]
@MarcoFalke removed from test framework in 8666f65
@fanquake additions in 375dcd1

**Review by ryanofsky — COMMENTED** [review]
Code review ACK 375dcd15566213c4d58e05a9738c8b5b1360a0d1. Various tweaks since last

**Comment by MarcoFalke** [comment]
I am thinking about merging this next week, but I am wondering if there are any outstanding questions or problems that need to be solved first?

File: CODEOWNERS
Comment (jonatack) [inline CODEOWNERS:15]: Should reviewers who wish to be added/removed make a PR for this? Should the process be described?

**Review by jonatack — COMMENTED** [review]
Concept ACK. Worth trying to see what the real effects will be. If it goes well, I could see more files or parts of the codebase added to the list.

Potential positives

- may stimulate review or accelerate it
- may provide info for newer people on who they can ask for help

Potential negatives

- may concentrate requests on fewer people
- the info for newer people will be incomplete if some don't wish to be on the list
- adds process (PRs to add/remove people?)

File: CODEOWNERS
Comment (jonatack) [reply CODEOWNERS:15]: Perhaps a gist or devwiki could be used to request being added/removed, and after a regular time period the changes to this file can be batched.

Comment (MarcoFalke) [reply CODEOWNERS:15]: I think anyone who has reviewed more than 5 pull requests AND has been active reviewing for more than 14 days should be allowed to add themselves in a pull request. Such a pull request doesn't need review and can be merged right away when the conditions have been fulfilled. (Exact conditions are up for debate)

**Comment by adamjonas** [comment]
Since tagged reviewers included in this file will receive notifications, I'd feel more comfortable if everyone listed opted-in before merging so that there aren't any surprises. Namely @sipsorcery, @practicalswift, @meshcollider, @achow101, and @sipa should have a chance to say no.

Reflected in 92f83731d76c0adc78f815aecac52eb0a7fcde91, I reached out to some folks out-of-band and made their requested adjustments. I also squashed the last commit since the concept has, for the most part, been ACKed.

**Review by hebasto — APPROVED** [review]
ACK 92f83731d76c0adc78f815aecac52eb0a7fcde91

**Comment by practicalswift** [comment]
ACK 92f83731d76c0adc78f815aecac52eb0a7fcde91 :)

**Comment by sipsorcery** [comment]
ACK 92f83731d76c0adc78f815aecac52eb0a7fcde91.

**Comment by MarcoFalke** [comment]
ACK 92f83731d76c0adc78f815aecac52eb0a7fcde91

**Comment by jonatack** [comment]
ACK 92f83731d76c0adc78f815aecac52eb0a7fcde91

**Comment by MarcoFalke** [comment]
@achow101 @meshcollider @sipa Progress here is currently blocked unless you select one of the following options:

* [ ] Remove your name for now
* [ ] Keep your name in for now

**Comment by jonatack** [comment]
@adamjonas If you have to retouch, feel free to add me to these files (if not too granular and at your discretion):

LOC touched:
- src/bitcoin-cli.cpp
- test/functional/feature_asmap.py
- test/functional/interface_bitcoin_cli.py
- test/functional/tool_wallet.py
-------
(I was tempted to propose these also, as I would not mind being notified, but would prefer to wait until others are added to them first. Posting this as a memo:
- src/net.h/cpp
- src/net_processing.h/cpp
- src/protocol.h/cpp
- src/txmempool.h/cpp
- src/validation.h/cpp
- src/validationinterface.h/cpp)

**Comment by luke-jr** [comment]
How does GitHub behave without this file? Does adding it lose some kind of automated suggestion of reviewers?

**Comment by adamjonas** [comment]
@luke-jr This file _adds_ automated reviewer suggestions. Without this file, the status quo remains the same.

**Comment by adamjonas** [comment]
[Touched-up](https://github.com/bitcoin/bitcoin/compare/92f8373..79a1d7708) with [additional files](https://github.com/bitcoin/bitcoin/pull/18949#issuecomment-635949562) for @jonatack and adding @achow101 to the wallet as discussed over IRC.

Still looking for the OK from @jonasschnelli, @sipa, and @meshcollider.

**Comment by fjahr** [comment]
If you have to retouch I would be interested in adding these:

```
# Coinstats
/src/node/coinstats.*	                    @fjahr

# Index
/src/index     			                    @fjahr
```

**Comment by meshcollider** [comment]
Hi sorry I missed the pings, I think this will make a lot of noise but I would be happy to give it a trial run at least, and remove it later if it is overwhelming. So concept ACK from me.

**Comment by adamjonas** [comment]
@meshcollider if it's all the same I'd rather remove you if you think it could lead to problems. I'm sure this file will require tweaks but I'd rather it adds more signal than noise. I've done the same for @jonasschnelli.

Other changes in 79042955f59bf56bc93aa93b6b69d3979a89b174 are adding @fjahr to the files he requested and @harding to the docs folder (excluding release notes).

**Comment by harding** [comment]
ACK (tested the pattern notifying me using `git ls-files`)

File: CODEOWNERS
Comment (sipa) [inline CODEOWNERS:46]: Add me for /doc/descriptors.md ?

File: CODEOWNERS
Comment (sipa) [inline CODEOWNERS:118]: Section for P2P handling? `/src/net_processing.*` and `/src/protocol.*`. Feel free to add me.

File: CODEOWNERS
Comment (sipa) [inline CODEOWNERS:125]: Other files: `/src/validation.*` `/src/{script,interpreter}.*` `/src/coins.*`.

File: CODEOWNERS
Comment (adamjonas) [reply CODEOWNERS:46]: I believe this should be covered by the [descriptors catch-all]( https://github.com/bitcoin/bitcoin/pull/18949/files#diff-5daf978e60e03b3975b485d5f1b449dcR104).

File: CODEOWNERS
Comment (adamjonas) [reply CODEOWNERS:137]: > /src/{script,interpreter}.*

I had some trouble matching this regex so I used `/src/script/interpreter.*` -- let me know if I got that wrong.

**Comment by jamesob** [comment]
Any chance I can jump on `coins.*`, `txdb.*`, and `dbwrapper.*` (unless it'll invalidate too many ACKs)?

File: CODEOWNERS
Comment (sipa) [reply CODEOWNERS:137]: I meant: group all of src/coins.* src/script/script.* src/script/interpreter.* src/validation.* and src/consensus/* under "consensus" - despite being spread out those are the most relevant files for defining the consensus implementation.

**Comment by jamesob** [comment]
Cool, thanks. ACK https://github.com/bitcoin/bitcoin/pull/18949/commits/2e2cf25ce98c4e1bca4647a2ab54cf5f38167703 from me.

**Comment by adamjonas** [comment]
Rebased, added sipa/ jameob, and removed redundant practicalswift line in the latest set of touch-ups.

**Comment by MarcoFalke** [comment]
re-crACK 2e2cf25ce98c4e1bca4647a2ab54cf5f38167703

File: CODEOWNERS
Comment (adamjonas) [reply CODEOWNERS:136]: Sorry for the slow response. I somehow missed this message. I've made the requested changes in the lastest push.

**Comment by adamjonas** [comment]
https://github.com/bitcoin/bitcoin/compare/2e2cf25..a06eb03 adds `src/script/script.*` and groups the consensus files as requested by @sipa.

**Comment by MarcoFalke** [comment]
I think this is ready to go in now. Should be trivial to edit/revert if it turns out to be not useful.

**Comment by adamjonas** [comment]
@MarcoFalke, do you think it makes sense to add a reference to this in CONTRIBUTING.md? I could also include the rules for inclusion that [you proposed](https://github.com/bitcoin/bitcoin/pull/18949/files#r429416244).

**Comment by MarcoFalke** [comment]
Before baking this into other documentation, let's try it out first and see whether it causes any issues.

**Pull request merged**
